lib/runtime-core/src/cache.rs

```diff
@@ -0,0 +1,181 @@
+pub enum Error {
+    IoError(io::Error),
+    DeserializeError(String),
+    SerializeError(String),
+    Unknown(String),
+    InvalidFile(InvalidFileType),
+    InvalidatedCache,
+}
+
+const CURRENT_CACHE_VERSION: u64 = 0;
+
+/// The header of a cache file.
+#[repr(C, packed)]
+struct CacheHeader {
+    magic: [u8; 8], // [W, A, S, M, E, R, \0, \0]
+    version: u64,
+    data_len: u64,
+    wasm_hash: [u8; 32], // Sha256 of the wasm in binary format.
+}
+
+impl CacheHeader {
+    pub fn read_from_slice(buffer: &[u8]) -> Result<(&CacheHeader, &[u8]), Error> {
+        if buffer.len() >= mem::size_of::<CacheHeader>() {
+            if &buffer[..8] == "WASMER\0\0".as_bytes() {
+                let (header_slice, body_slice) = buffer.split_at(mem::size_of::<CacheHeader>());
+                let header = unsafe { &*(header_slice.as_ptr() as *const CacheHeader) };
+
+                if header.version == CURRENT_CACHE_VERSION {
+                    Ok((header, body_slice))
+                } else {
+                    Err(Error::InvalidatedCache)
+                }
+            } else {
+                Err(Error::InvalidFile(InvalidFileType::InvalidMagic))
+            }
+        } else {
+            Err(Error::InvalidFile(InvalidFileType::InvalidSize))
+        }
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        let ptr = self as *const CacheHeader as *const u8;
+        unsafe { slice::from_raw_parts(ptr, mem::size_of::<CacheHeader>()) }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct CacheInner {
+    info: Box<ModuleInfo>,
+    #[serde(with = "serde_bytes")]
+    backend_metadata: Vec<u8>,
+    compiled_code: Memory,
+}
+
+pub struct Cache {
+    inner: CacheInner,
+    wasm_hash: Box<[u8; 32]>,
+}
+
+impl Cache {
+    pub(crate) fn new(
+        wasm: &[u8],
+        info: Box<ModuleInfo>,
+        backend_metadata: Vec<u8>,
+        compiled_code: Memory,
+    ) -> Self {
+        let wasm_hash = hash_data(wasm);
+
+        Self {
+            inner: CacheInner {
+                info,
+                backend_metadata,
+                compiled_code,
+            },
+            wasm_hash: Box::new(wasm_hash),
+        }
+    }
+
+    pub fn open<P>(path: P) -> Result<Cache, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let file = File::open(path).map_err(|e| Error::IoError(e))?;
+
+        let mmap = unsafe { Mmap::map(&file).map_err(|e| Error::IoError(e))? };
+
+        let (header, body_slice) = CacheHeader::read_from_slice(&mmap[..])?;
+
+        let inner =
+            deserialize(body_slice).map_err(|e| Error::DeserializeError(format!("{:#?}", e)))?;
+
+        Ok(Cache {
+            inner,
+            wasm_hash: Box::new(header.wasm_hash),
+        })
+    }
+
+    pub fn info(&self) -> &ModuleInfo {
+        &self.inner.info
+    }
+
+    pub fn wasm_hash(&self) -> &[u8; 32] {
+        &self.wasm_hash
+    }
+
+    #[doc(hidden)]
+    pub fn consume(self) -> (ModuleInfo, Vec<u8>, Memory) {
```

Comment:
What this function is for?
Can we make it public only for the crate?

Reply:
Not really, since it's used in the clif backend.